### ISSUE TITLE
Failed to allocate memory to create Connection

### ISSUE THREAD

**Issue by Arohdin** [Azure/azure-event-hubs-python#43]
Hi

I am currently upgrading from sdk version 0.2.0b2 to version 0.2.0rc1; using python 3.5.
I've gone ahead and changed my code according to the example [Receiving from IoTHub asynchronously with multiple receivers](https://github.com/Azure/azure-event-hubs-python/blob/master/tests/test_iothub_receive_async.py) given by **annatisch** from [issue #28](https://github.com/Azure/azure-event-hubs-python/issues/28).

Whenever I run my updated code I get the error: `Failed to allocate memory to create Connection`.
This error seems to come from within the `<EventHubClientAsync>.run_async()` function.

I can't really supply source code here due to policies; but here is what I am doing.

Main python script creates a new thread using `Threading`.
In the new thread I run a function/script that does some things.
1. Create and Set new event loop in asyncio.
2. Creates a new `EventHubClientAsync` object
3. Retrieves all partitions for a given eventhub.
4. Creates a list and fill it with async receivers using `<EventHubClientAsync>.add_async_receiver()`.
5. runs `<EventHubClientAsync>.run_async()`.
6. ....
.
.
.

It is at step 5 that the program returns the given error.

I am currently on Linux, using an Ubuntu 16.04 based distro called Elementary OS Loki 0.4.
If that's of any help.

Thanks

_EDIT:_ _I am upgrading eventhub sdk in a multipart effort to fix a problem I got when using 0.2.0b2. The problem there was that a long running script would crash with error: `Segmentation Fault (Core Dumped)`_
[Just noticed this ticket; it's still a problem for me at version 0.2.0b2](https://github.com/Azure/azure-event-hubs-python/issues/10)

**Comment by annatisch** [comment]
Thanks @Arohdin - The connection object is created from parameters mostly derived from how you instantiate the `EventHubClientAsync`.
Are you able to show me just the code that creates the client? And also some variation on the connection string you're passing in to create the client (with sensitive info swapped out, but structurally the same).
Additionally - are you running multiple clients across multiple threads? Or does this occur when only one client is running?

Thanks!

**Comment by annatisch** [comment]
hmm actually in addition to my previous question - I see that I connection must be successfully created in order to retrieve the EventHub info like partitions. Would it also be possible to see a couple of lines of code where your retrieve the partitions?

**Comment by Arohdin** [comment]
Thanks for the quick response @annatisch 

> The connection object is created from parameters mostly derived from how you instantiate the EventHubClientAsync.
Are you able to show me just the code that creates the client?

This is how I currently create the `EventHubClientAsync` instance.
`client = EventHubClientAsync(ADDRESS, debug=False, username=USER, password=KEY)`

> are you running multiple clients across multiple threads? Or does this occur when only one client is running?

I am currently only running one client on one thread.

> Would it also be possible to see a couple of lines of code where your retrieve the partitions?

` for i in range(len(partitions)):`
`rcvrs.append(client.add_async_receiver(CONSUMER_GROUP, partitions[i], OFFSET, 0))`


In addition, did you see the _EDIT_ referencing issue #10?

**Comment by Arohdin** [comment]
Thanks for the quick response @annatisch 

> The connection object is created from parameters mostly derived from how you instantiate the EventHubClientAsync.
Are you able to show me just the code that creates the client?

This is how I currently create the `EventHubClientAsync` instance.
`client = EventHubClientAsync(ADDRESS, debug=False, username=USER, password=KEY)`

> are you running multiple clients across multiple threads? Or does this occur when only one client is running?

I am currently only running one client on one thread.

> Would it also be possible to see a couple of lines of code where your retrieve the partitions?

In order to retrieve the partitions I've got a function which returns: `await client.get_eventhub_info_async().get('partition_ids')`. 
This is a list with all partition ids.


In addition, did you see the _EDIT_ referencing issue #10?

**Comment by Arohdin** [comment]
Okey!

I actually just used EventHub specific `EventHubClientAsync` due to it working. And yes, I am using the `Event Hub-compatible endpoint`.

I just now tried to change it to `EventHubClientAsync.from_iothub_connection_string` but I can't get it to work. I get `Unable to open management session. Please confirm URI namespace exists.`
Which connection string am I suppose to supply?

**Comment by annatisch** [comment]
Thanks - what happens if you remove the call to retrieve the partition IDs for now and just hardcode in a list like `["0", "1"]` - just to see whether the rest of the client works?

The IotHub connection string I use from the portal is from the standard IoThub settings with the `iothubowner` policy.

**Comment by Arohdin** [comment]
Connected usings normal `EventHubClientAsync` and then hardcoded the partitions (1-4) and still got the same error: `Failed to allocate memory to create Connection`.

I accidentally used the wrong connection string when trying `EventHubClientAsync.from_iothub_connection_string`, sorry.
Supplying the correct one (from SETTINGS -> Shared access policies -> iothubowner ....) gives the error: `Failed to open mgmt link: MgmtOpenStatus.Error`

**Comment by annatisch** [comment]
Thanks! I think the only part of the code where it will raise `Failed to open mgmt link` is trying to get the partition IDs - so it must still be calling `get_eventhub_info(_async)` somewhere?

**Comment by annatisch** [comment]
Thanks for these details - I will run some more tests to see what I can discover.

When you say that sometimes only 1 or 2 partitions failed - does that mean that the remaining open partitions successfully received events?

**Comment by Arohdin** [comment]
No. The one that passes `client.run_async()` instead gets an error when it tries to receive using the `.receive()` function in the next part.

**Comment by annatisch** [comment]
Gotcha - so all the receivers eventually fail with the same error - just that the timing varies.
I will investigate - thanks for the report!

**Comment by Arohdin** [comment]
@annatisch 

**An update for the segmentation fault problem.
I ran 'gdb' to see where the c library crashed and I got this result.**
`AttributeError: 'ConnectionAsync' object has no attribute '_io_error'
Exception ignored in: 'uamqp.c_uamqp.on_io_error'
AttributeError: 'ConnectionAsync' object has no attribute '_io_error'
Failed to open receiver: Message Receiver Client was unable to open. Please confirm credentials and access permissions.
See debug trace for more details.
Please check that the partition key is valid and whether an epoch receiver is already running for this partition.
`


**I ran the backtrace command and got this:**
``` 
#0  0x0000000000000330 in ?? ()


#1  0x00007ffff316fa3d in on_bytes_encoded ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so


#2  0x00007ffff31742a4 in frame_codec_encode_frame ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#3  0x00007ffff31811c7 in amqp_frame_codec_encode_frame ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#4  0x00007ffff31733d5 in connection_encode_frame ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#5  0x00007ffff31806ed in session_send_detach ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#6  0x00007ffff317462b in send_detach.isra.3.constprop ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#7  0x00007ffff317635d in link_detach ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#8  0x00007ffff317651e in link_destroy ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#9  0x00007ffff30c4386 in __pyx_f_5uamqp_7c_uamqp_5cLink_destroy (
    __pyx_v_self=0x7ffff2a9b670, __pyx_skip_dispatch=0) at uamqp/c_uamqp.c:45590

#10 0x00007ffff30c2969 in __pyx_pf_5uamqp_7c_uamqp_5cLink_2__dealloc__ (
    __pyx_v_self=0x7ffff2a9b670) at uamqp/c_uamqp.c:45172

#11 0x00007ffff30c1e51 in __pyx_pw_5uamqp_7c_uamqp_5cLink_3__dealloc__ (
    __pyx_v_self=0x7ffff2a9b670) at uamqp/c_uamqp.c:45035

#12 0x00007ffff31259d4 in __pyx_tp_dealloc_5uamqp_7c_uamqp_cLink (o=0x7ffff2a9b670)
    at uamqp/c_uamqp.c:81733

#13 0x00000000005a1015 in ?? ()
#14 0x00000000004acd94 in ?? ()

#15 0x00007ffff3040222 in __Pyx_PyObject_GetAttrStr (obj=0x7ffff2a9ae10, 
    attr_name=0x7ffff3aaabf0) at uamqp/c_uamqp.c:2495

#16 0x00007ffff314c182 in __Pyx_GetAttr (o=0x7ffff2a9ae10, n=0x7ffff3aaabf0)
    at uamqp/c_uamqp.c:93058

#17 0x00007ffff314c1fe in __Pyx_HasAttr (o=0x7ffff2a9ae10, n=0x7ffff3aaabf0)
    at uamqp/c_uamqp.c:93071

#18 0x00007ffff30e4adc in __pyx_f_5uamqp_7c_uamqp_on_message_receiver_state_changed (
    __pyx_v_context=0x7ffff2a9ae10, __pyx_v_new_state=MESSAGE_RECEIVER_STATE_CLOSING, 
    __pyx_v_previous_state=MESSAGE_RECEIVER_STATE_OPEN) at uamqp/c_uamqp.c:55951

#19 0x00007ffff317972b in messagereceiver_close ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#20 0x00007ffff31797fe in messagereceiver_destroy ()
   from /home/arohdin/Git/virtualenvironments/summerjobIoT/lib/python3.5/site-packages/uamqp/c_uamqp.cpython-35m-x86_64-linux-gnu.so

#21 0x00007ffff30e3519 in __pyx_f_5uamqp_7c_uamqp_16cMessageReceiver_destroy (
    __pyx_v_self=0x7ffff2a9b690, __pyx_skip_dispatch=0) at uamqp/c_uamqp.c:55489

#22 0x00007ffff30e0567 in __pyx_pf_5uamqp_7c_uamqp_16cMessageReceiver_2__dealloc__ (
    __pyx_v_self=0x7ffff2a9b690) at uamqp/c_uamqp.c:54773

#23 0x00007ffff30dfa4f in __pyx_pw_5uamqp_7c_uamqp_16cMessageReceiver_3__dealloc__ (
    __pyx_v_self=0x7ffff2a9b690) at uamqp/c_uamqp.c:54636

#24 0x00007ffff31263b9 in __pyx_tp_dealloc_5uamqp_7c_uamqp_cMessageReceiver (
    o=0x7ffff2a9b690) at uamqp/c_uamqp.c:82217

#25 0x00000000005813aa in ?? ()
#26 0x0000000000581989 in ?? ()
#27 0x000000000051860d in ?? ()
#28 0x000000000060a2cd in ?? ()
#29 0x000000000060a32e in PyGC_Collect ()
#30 0x000000000060da09 in Py_Finalize ()
#31 0x000000000063fd16 in Py_Main ()
#32 0x00000000004cfeb1 in main ()
```


**This happens after the has been running for some time. It seems to die when at the moment when I try to connect, although I am not sure. Quite hard to reproduce the crash at will.**

**Please let me know if this is of interest.**

**Comment by annatisch** [comment]
Thanks @Arohdin - just to confirm - this crash is happening independently from the "did not match any of the expected formats" error?

**Comment by Arohdin** [comment]
@annatisch Yes. The segmentation fault crash is a problem that I have using 0.2.0b2.
Sorry for posting about two separate issues in one thread. Should I create a new one instead?

**Comment by annatisch** [comment]
Thanks for your patience @Arohdin - no need to publish another issue for the b2 crash.

Status update:
I've found the root of and fixed the "Failed to allocate memory" error itself.
However it seems that I'm still getting errors when attempting to retrieve the eventhub info except now in the form of:
```
2018-07-19 17:06:07,967 uamqp.c_uamqp INFO     b"<- [DETACH]* {2,true,* {amqp:unauthorized-access,Unauthorized access. 'Listen' claim(s) are required to perform this operation. Resource: 'sb://iothub-ns-pyot-xxxxxx-8fa45cb72e.servicebus.windows.net/pyot/$management'. TrackingId:565e6f7201f9479588b0fc8620ab84b6_G17, SystemTracker:gateway5, Timestamp:7/20/2018 12:06:07 AM,NULL}}"
2018-07-19 17:06:07,967 uamqp.c_uamqp INFO     b"<- [DETACH]* {3,true,* {amqp:unauthorized-access,Unauthorized access. 'Send' claim(s) are required to perform this operation. Resource: 'sb://iothub-ns-pyot-xxxxxx-8fa45cb72e.servicebus.windows.net/pyot/$management'. TrackingId:565e6f7201f9479588b0fc8620ab84b6_G17, SystemTracker:gateway5, Timestamp:7/20/2018 12:06:07 AM,NULL}}"
```
I haven't yet figured out whether this is simply a result of my IotHub setup, or whether it's a bigger issue. Regardless I will try to publish a release with the partial fix early next week.

**Comment by annatisch** [comment]
Okay I found the source of the unauthorized-access errors (yes, it was a bug, not my setup) and have resolved that as well. Aiming to have a uAMQP release out today or tomorrow and an EH release will follow soon after (sometime this week).

Thanks for your patience!

**Comment by Arohdin** [comment]
@annatisch Great to hear! Will get back to you when I get a chance to test the new version.

**Comment by annatisch** [comment]
Thanks @Arohdin - this is now published.

The process for working with IoTHub is a bit different from a standard EventHub. At least one receiver must attempt to connect and be redirected by the service in order to get the correct endpoint information. And it's not until we have this correct endpoint information that we can establish the management link to get the EventHub info like PartitionIds. But we don't know how many receivers to add until we know the partitions - so a bit of a chicken and egg scenario.

I would recommend doing the approach I've used in my test here - where an independent client is created with a single receiver just to collect the partition information. That is then used to set up the main client for receiving purposes:
https://github.com/Azure/azure-event-hubs-python/blob/v0.2.0rc2/tests/test_iothub_receive_async.py

Let me know how it goes :)

**Comment by Arohdin** [comment]
Thanks @annatisch 

I've got the new version up and running. Currently doing a longer testing (couple of days).
I'll get back to you when I know if everything works as intended.

Cheers

**Comment by annatisch** [comment]
Hi @Arohdin - how's this going?
I've just published a new release of both uAMQP (0.2.1) and azure-eventhub (0.2.0) with some further stability improvements.

If this issue seems to be resolved I can close it - if you encounter any new problems, we can open fresh issues :)

**Comment by Arohdin** [comment]
@annatisch 

Sorry for the delay, I somewhat forgot about this thread (woops).
I am currently on 0.2.0rc2 and it works as intended without any issues and thus you can close this thread.

Thanks for all the help :smile: :+1: 

Ps. I probably won't be upgrading to 0.2.0,  unless there is any major stability improvements, due to the project coming to an end (at least for me).

**Comment by annatisch** [comment]
Good to hear :)
